### PR TITLE
Support for exceptions with detailed error information

### PR DESCRIPTION
**Pull request by Fluxie**
I decided not to introduce additional hierarchy and inherited the exceptions directly from appropriate std exceptions instead.

This resolves issue #45.

### CODE REVIEW

File: intercom-cpp/src/detail/utility.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+     * @return false The machine is big endian.
+     */
+    inline bool is_little_endian()
+    {
+        // Create a copy of the data with memcpy.
+        // Directly casting with e.g. reinterpret_cast will lead to UB.
+        int test_value = 1;
+        std::array< uint8_t, sizeof( int ) > asBytes;
+        std::memcpy( &asBytes, &test_value, sizeof( int ) );
+        return asBytes[ 0 ] == 1;
```
Comment (Rantanen) [inline intercom-cpp/src/detail/utility.h]: `*reinterpret_cast<*char>(..)` should not result in UB. It's one of the few exceptions. Note that `uint8_t` and `char` are different types, so while they have the same size, `uint8_t` reinterpret cast is still UB.

Of course on the other hand, `char` might not be 8-bit type, but at that point `uint8_t` wouldn't exist either as `char` is required to be the smallest type as far as I've understood.

So while using `char` for `is_little_endian` checks, using it for the actual hex-formatting might not be fully cross platform.

File: intercom-cpp/src/detail/utility.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        stream << std::hex << std::uppercase;
+        if( is_little_endian() )
+        {
+            // On little-endian machine the bytes are on reverse order.
+            for( auto byte = asBytes.rbegin(); byte != asBytes.rend(); ++byte )
+            {
+                // Explicit cast required, otherwise byte is treated as unsigned char.
+                stream << static_cast< unsigned int >( ( *byte & 0xF0 ) >> 4 );
+                stream << static_cast< unsigned int >( *byte & 0x0F );
+            }
```
Comment (Rantanen) [inline intercom-cpp/src/detail/utility.h]: Bytes of _individual fields_ are on reverse order. The fields themselves are in the same order. So if `TData` is a struct, this doesn't apply.

Namely with `struct S { int a; int b }`, `&s.a` should always be the start of the struct, with `&s.b > &s.a`.

In the use case (like in the PR) where `TData` is an integral type, this isn't a problem. However if this appears in the public API of the C++ library, we might want to hide the template implementation behind an anonymous namespace and add overloaded `write_as_hex` implementations for the compatible (ie. integral) types.

Anyway that can be done at a later point. I think there's also some similar API work to be done in Rust-side.

**Comment by Rantanen** [comment]
Few small nits but <3 exceptions, so I don't really care for them now. Thanks!

File: intercom-cpp/src/detail/utility.h (hunk 2 above)
Comment (Fluxie) [reply intercom-cpp/src/detail/utility.h]: "write_as_hex" is a "detail" of intercom library which is supposed to be private by convention. I want to use it from different places within intercom so I really cannot put it inside anonymous namespace.

Other than unsigned integral types can be excluded with a bit appropriate [template](http://www.cplusplus.com/reference/type_traits/is_unsigned/). I'll implement that soonish.

**Pull request merged**